### ISSUE TITLE
reportURL block does not work when accessing Snap! over https

### ISSUE THREAD

**Issue by thomaswp** [jmoenig/Snap--Build-Your-Own-Blocks#1762]
If someone wants to use Snap! over `https` (https://snap.berkeley.edu/run), for example to include it in an iframe on a secure webpage, or because they generally like to practice safe internet usage, the `reportURL` block does not work. 

![image](https://user-images.githubusercontent.com/1750176/27086596-a6fe9b38-5020-11e7-9355-cc0c969485b0.png)

This is the case for obvious reasons: the block always uses `http` and modern browsers will not allow insecure AJAX requests on a secure page.

I've submitted a simple fix that simply uses `document.location.protocol` instead of a hardcoded `http:`. A more robust solution would probably give the `reportURL` block a dropdown of protocol choices.

Obviously if the page you want to fetch doesn't support https, you can't get the data when using Snap over http no matter what you do, so there's only so much we can do.

**Comment by cycomachead** [comment]
Are there many pages that _don't_ support https? I kind of think we should just change the fault for every request, but I don't know how likely this is to break anything.

**Comment by liam4** [comment]
> Are there many pages that don't support https?

Typically locally hosted servers use HTTP, not HTTPS – for example you might make a simple Python program that writes files to disk when it gets HTTP requests; it would be inconvenient to have to program in HTTPS support for a simple experiment server like that.

**Comment by jguille2** [comment]
Hi everybody!
I understand this issue (_http_ requests are not allowed under an "_https_ environment") but I think we need a more generic solution (also allowing _https_ requests under a typical "_http_ snap").

Then Yes, we need a dropdown of protocol options.
![blockshttp](https://user-images.githubusercontent.com/2926315/27150819-fa067afc-5148-11e7-99ae-a7269623abad.png)
![httpoptions](https://user-images.githubusercontent.com/2926315/27150827-005e8778-5149-11e7-8217-90e3afe188c1.png)
 It is an easy implementation... but it need to be blessed, because this question was opened (and not closed) in the past (for example [here](https://github.com/jmoenig/Snap--Build-Your-Own-Blocks/issues/782))

In the meantime, you can build your own custom block to do this job. If you want, I share [this project](http://snap.berkeley.edu/snapsource/snap.html#present:Username=jguille2&ProjectName=httpRequest) with a functional http/https reporter.
![blockhttps](https://user-images.githubusercontent.com/2926315/27150961-61b6b252-5149-11e7-8d5a-d0822681109b.png)

Joan

**Comment by brianharvey** [comment]
Is this different from what's in the Web library in Snap<i>!</i>?

**Comment by cycomachead** [comment]
As long as we don't ban JS Functions any time soon, 🙂, this is all possible with the web requests library. 

My question ( and I believe Thomas' motivation, too) are about easy to use defaults. 

IMO in 2017 we should default to https both in terms of what s.b.e serves and what the http block does. 

Clearly we need to support both protocols - I guess that means a built in block with a drop down...though I'm not in love with that. A drop down is more complexity than needed 90%+ of the time. 

My ideal solution would be to migrate the built in block to https and then to dynamically import a fallback http block for existing projects. 
New Projects that need to make requests to local servers can do so using the web requests library. 

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by jguille2** [comment]
What would I do if JS functions were banned? For me, it is one of the most important features!

Yes, Brian, I can do it with the http lib... but I think it's no so obvious (hide header input...)
![httpblocklib](https://user-images.githubusercontent.com/2926315/27155312-c4c7c6d8-5158-11e7-92d2-d5589730d441.png)

I don't think the dropdown options is so complex... In fact, it is the same current block (the 'http' default) with the option to change to 'https'.

Another option is an 'url reporter block', writing the protocol into the url input (just like the http lib block).
![urlblocks](https://user-images.githubusercontent.com/2926315/27155693-310a1c96-515a-11e7-87fa-1027e8065e8e.png)
In fact, copying the url from the browser or from a link, we are copying this full text.
![snapurl](https://user-images.githubusercontent.com/2926315/27155969-54eb1eca-515b-11e7-9291-4a29fc5268b5.png) --> http://snap.berkeley.edu/

And with any option, we do not have to worry about backward compatibility...with _blockMigrations_ we can arrange and adapt the changes (with the dropdown  `inputs: [['http'], '<#1>']` or with the url block `inputs: [['http'] + '<#1>']`

**Comment by cycomachead** [comment]
We're not actually gonna kill the JS function anytime soon, but #1754.

Also, in any event I would advocate we have a couple more blocks in the web requests library that are a http request with varying degrees of complexity.

> I don't think the dropdown options is so complex... In fact, it is the same current block (the 'http' default) with the option to change to 'https'.

Every single option adds some cognitive overhead and gives more room for confusion. A dropdown isn't a bad solution, but it's a cop-out. My guess is that 90% (maybe even more) of the http requests made would work fine as https requests. The issues are those made to local servers - which I think aren't a big problem because in most of those projects (like the Finch, WaterColorBot, and whatnot) students are interacting through a library not actual http blocks. 

I like the `url:` solution since it does make C&P easy for students, though I still favor including the protocol because I think it's a little bit more clear what's going on. (If you're not aware of what a "URL" is, you  can still match `http(s)` and guess at what's going on.)

> And with any option, we do not have to worry about backward compatibility...with blockMigrations we can arrange and adapt the changes (with the dropdown inputs: [['http'], '<#1>'] or with the url block inputs: [['http'] + '<#1>']

Assuming we go with a one block solution - then yes, of course it's possible. My point was that it's tricky to exactly what I think is best. (i.e. One block with no dropdown.) That would require blockMigrations to support loading a new or hidden block - not just translating parameters. 
Actually, if we kept the source for reportURL we could keep the http block around, but not in the palette and just have a separate https block.

**Comment by brianharvey** [comment]
> I like the url: solution since it does make C&P easy for students, though I still favor including the protocol because I think it's a little bit more clear what's going on.

If the "get URL" block had a default input of "https://snap.berkeley.edu" that would remind everyone what a URL is.

We could also (should, I think) still call it HTTP:// but accept full URLs as input and do the right thing.  That would avoid needing a special backward compatibility block.

**Comment by cycomachead** [comment]
> We could also (should, I think) still call it HTTP:// but accept full URLs as input and do the right thing. That would avoid needing a special backward compatibility block.

While I agree we should do the right thing - I don't think this _avoids_ fixing the problem / adapting / creating a block. If you need to make an https request, there's no way that it makes sense for a block that says "http://" to let you do this. Also, with existing blockMigrations code, it's really pretty easy to migrate to a `url:` block.

There's something about a url block that doesn't feel quite as nice as the current block, but it does seem to be an improvement over the current situation and better than a dropdown or multiple blocks.

**Comment by bromagosa** [comment]
We should indeed serve https by default. The new API won't work otherwise, neither will the new webcam dialog.

**Comment by jguille2** [comment]
OK,
I agree wit @brianharvey 
![geturl](https://user-images.githubusercontent.com/2926315/27171383-99b52360-51b1-11e7-8acb-1dfe08e3f2b8.png)
It is fine and we don't need _blockMigrations_

This runs ok, and we can use any protocol (http, https, file...) and if the input has not the protocol, it uses the default one.
```javascript
Process.prototype.reportURL = function (url) {
    var response;
    if (!this.httpRequest) {
        this.httpRequest = new XMLHttpRequest();
        if (url.indexOf('://') == -1) {
            this.httpRequest.open("GET", 'http://' + url, true);
        } else {
            this.httpRequest.open("GET", url, true);
        }
        this.httpRequest.send(null);
    } else if (this.httpRequest.readyState === 4) {
        response = this.httpRequest.responseText;
        this.httpRequest = null;
        return response;
    }
    this.pushContext('doYield');
    this.pushContext();
};
```
**Notes:** The default beaviour may be:
  -  _http_ (I vote for this, for full backward compatibility),
  -  or _https_ (as Michel said, " 90% (maybe even more) of the http requests made would work fine as https requests"),
  - or I can use _document.location.protocol_ (as the @thomaswp proposal, to use dinamically 'current snap protocol used').

**Comment by cycomachead** [comment]
We can't default the protocol to https without breaking some projects. :( However, I do think https is the right default. Given that courses like BJC also talk about the importance of security (on principle and for practical purposes), I think it's also important for us to follow.

I don't really like defaulting to a variable that can change. You'd get different behavior with no clear reason why. If something were to break, it would be incredibly difficult to debug. 

There are 2 solutions that could work, I think:
* The Snap<em>!</em> XML format has a version attribute. We could increment the version and change the default protocol for older versions, or (more likely) prepend the protocol to the address of the http block - but that involves inserting code into projects.
* Maintain 2 blocks. The `http://` block and a new `URL:` block. The code and block definition for the `http://` could exist in Snap<em>!</em>, but never added to the palette. Thus old projects load fine, and can use / execute the "old" block, but every project now has access to the newer block.
There's a bit of oddity where when you open an old project you'd have both the `URL:` and the `http://` blocks in use, with slightly different behavior - but I don't think this would affect very many users.

**Comment by jguille2** [comment]
Hi Michael,
In my opinion... The simpler proposal is not wrong. The idea is that _url_ param, always had the protocol (the default is only for backward compatibility with _http://_ block). If you want to explain to the students the importance of security, they have to make sure to write _https_ in their addresses (like in browsers).

In any case I think the alternative wouldn't be two blocks. In this case, we can use blockMigrations to change
![anet](https://user-images.githubusercontent.com/2926315/27173769-0e58cae8-51ba-11e7-8f1d-9f5dbea96424.png)
to
![urlanet](https://user-images.githubusercontent.com/2926315/27173771-12e57548-51ba-11e7-90de-ff70eed9a470.png)
I must check the possibilities of playing with <#1> to be sure.

**Comment by cycomachead** [comment]
We can't migrate that because there could be another block in there - we could use `join`, which is what I was referencing with modifying the code. It is possible to do the migration, regardless - it just becomes a question of how complex. That's why I suggested two blocks because it's very simple. (And critically, most people won't notice there are 2 blocks - so we aren't introducing much complexity.)

> The idea is that url param, always had the protocol (the default is only for backward compatibility with http:// block). If you want to explain to the students the importance of security, they have to make sure to write https in their addresses (like in browsers).

I guess this is why I like the protocol-less version better. Protocols are a technical thing most people shouldn't need to worry about. I very fundamentally disagree with the idea that if people want security they should do it themselves. Every major site redirects users to https by default now, and browsers autocomplete domains. I never type `http(s)`, except when programming. By having a more secure default, we do everyone a service because we (more so) protect that traffic from being snooped on. We also make it easier to use sites that only work with https. This is at the expense of sites that only work with http, which are slightly more effort now - but I don't expect many students to directly run into such issues (since students who use locally hosted servers are often using them through a nicely structured library).

I also think http as a default doesn't make sense because at some point we will serve s.b.e from an https domain - and in that case we _will_ need to change the default, otherwise the vast majority of existing projects will break.

I guess in the case of serving from https, a solution that permanently migrates existing protocol-less code to http (not https) requests would actually make our lives more difficult in the future - because now they'll fail and we can't know if we should change the URL.

Given that, it seems like we'll need an https default at some point...

There are two ideas I have now, but both seem complicated.
* we could detect https failures and then try a second request with an http url
* we can specifically treat protocol-less requests that resolve to a localhost (or local IP) as using http. This isn't so bad, as there's only a handful of addresses that are used for local development. (And yes, people can edit their own hosts to use custom domains, but I really doubt and Snap<em>!</em> students are doing this...)

**Comment by cycomachead** [comment]
A discussion that probably needs to be separated, but: @thomaswp I'm assuming this is for edX PD? 
If so, are you relying specifically on s.b.e over https?

In that case, I think the PR is a good stop-gap, because most projects will work fine and nearly everyone will be none-the-wiser about subtle differences behind the scenes.

**Comment by brianharvey** [comment]
I'm thinking about the block name...  GET URL may make a broader promise to the user than we are actually going to carry out.  URLs include file: and mailto: and other such things.  Maybe it should instead be GET WEB PAGE or something?

What would be truly wonderful is if browsers had "virtual tabs" so that we could pass the URL to the browser, let it do all its magic, and then scrape the actual displayed text instead of seeing a bunch of JS and CSS.

Back to reality, I'm in favor of us serving https by default sooner rather than later.  What are the implications for interaction with Miosoft?  And I'm also in favor of fallback to http if https fails, for the GET WEB PAGE or whatever.

**Comment by cycomachead** [comment]
I don't want to use the word "web page" because this works for APIs and in the future I imagine images and the like. (Also in terms of privacy and security, it is pretty darn good that we can't just fully render and interact with a page in another tab.)

As far as https on s.b.e: no implications for Miosoft. I'll have to see how local storage treats the different domains. 

I wanted to list the pros and cons for each option - because I think it will help us think about the users we are trying to help most. 

* block with a drop down:
Pros:
   - easy to implement
   - easy for users to switch protocols if something goes wrong
   - easy for us to change the default in the future without affecting projects using the block. 
Cons:
   - I think an option is distracting nearly all the time. And can lead to unnecessary questions.
   - we still need to choose a default to migrate existing blocks to. 
Generic Block with http default:
 Pros: 
   - simple to implement too
   - minimal current changes
Cons:
   - less secure and private mechanism for 95% of requests that work over both protocols
   - doesn't really help us for when we move to an https domain
    - will make it more difficult to change in the future (or at least we have the same problems we are currently discussing)

Default to https:
Pros:
   - secure by default
   - works regardless of domain
Cons:
   - deciding how to migrate existing projects is tricky.
   - likely some small amount of broken requests (but we should be able to have a fallback)
   - most technically complex option 

-----

So that's for the 3 blocks assuming we stay on an http domain with the intent to migrate to https in the future. 

I'm realizing that given the actual problem of failing requests from https pages, even if we protect local host requests to be http, we will probably still run into security issues. :( I'll have to see if there's a way to work around that.  

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by thomaswp** [comment]
Wow, I didn't expect this to trigger such a good discussion!

I'm pretty happy with any of the proposed solutions, and I think @cycomachead enumerated the pros and cons. I would argue that the "con" of the http/https dropdown adding overhead and confusion is overstated, since most users will never bother trying to change it. It could actually be a "pro," since it gives teachers an opportunity to talk about internet security, which I believe is relevant in BJC.

If we go the dropdown route, I'd suggest defaulting to http:// for imported projects to preserve backwards compatibility, but default to https:// for new projects. 

@cycomachead, as for my needs with the BJC PD and edX (which was the original impetus for this issue), I'm using my own fork, and I've already used the code in my pull request, so no rush on my account.

**Comment by jguille2** [comment]
Hi,
I think the 'generic block with default https' is the best option.

I would like to have a new block (I think the name 'get URL:' is not bad) and arrange backward compatibility with blockMigrations. But I can't. I don't know how to load the new block changing the "host.dom/opt" input  to "http://host.dom/opt".

Maybe we can implement new features to _blockMigrations_ ... but it's not clear for me (in 'dev' we have an 'offset' option... but this is not enough and we would need operations with <#1> to do this).

Then I think the remaining option is (as @cycomachead said) to keep the old "http://" block in the background. This block will not be into the palette, but it will appear loading old projects. And I think the situation will be clear: the new block uses 'https' by default, but it invites you to explicit the scheme into the url (as @brianharvey said "If the "get URL" block had a default input of "https://snap.berkeley.edu" that would remind everyone what a URL is.")

**Comment by cycomachead** [comment]
Fixed with 37dbcad27f03fe9edd8c43875305c3d2c29b59f9
